components/authz-service/server/v2/rules_property_test.go

```diff
@@ -460,7 +463,108 @@ func TestDeleteRuleProperties(t *testing.T) {
 			},
-			createProjectAndRuleGen,
+			createProjectAndRulesGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestListProjectProperties(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+
+	properties.Property("all staged rules are reported as staged",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				reqs.rules, err = createRules(ctx, cl, reqs.rules)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				resp, err := cl.ListProjects(ctx, &api.ListProjectsReq{})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				if 1 != len(resp.Projects) {
+					return false
+				}
+				respProject := resp.Projects[0]
+				t.Logf("===> %d rules", len(reqs.rules))
+				return respProject.Id == reqs.CreateProjectReq.Id &&
+					respProject.Name == reqs.CreateProjectReq.Name &&
+					respProject.Type.String() == "CUSTOM" &&
+					respProject.Status == storage.EditsPending.String()
+			},
+			createProjectAndRulesGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestListProjectPropertiesMixingStagedAndApplied(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+
+	properties.Property("projects with staged, applied, or no rules are reported",
+		prop.ForAll(
+			func(reqs []projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+				t.Logf("LIST %d PROJECTS ===>", len(reqs))
+				applyPoint := rand.Intn(len(reqs))
+
+				for i, req := range reqs {
+					if applyPoint == i {
+						t.Log("*** applying rules now ***")
+						cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+					}
+					t.Logf("===> project '%s' has %d rules", req.CreateProjectReq.Id, len(req.rules))
+					_, err := cl.CreateProject(ctx, &req.CreateProjectReq)
+					if err != nil {
+						return reportErrorAndYieldFalse(t, err)
+					}
+					req.rules, err = createRules(ctx, cl, req.rules)
+					if err != nil {
+						return reportErrorAndYieldFalse(t, err)
+					}
+				}
+
+				resp, err := cl.ListProjects(ctx, &api.ListProjectsReq{})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				if len(reqs) != len(resp.Projects) {
+					return false
+				}
+
+				for i, p := range resp.Projects {
+					if len(reqs[i].rules) == 0 && p.Status != storage.NoRules.String() {
+						return false
+					} else if len(reqs[i].rules) > 0 {
+						if i < applyPoint {
+							if p.Status != storage.Applied.String() {
+								return false
+							}
+						} else {
+							if p.Status != storage.EditsPending.String() {
+								return false
+							}
+						}
+					}
+				}
+				return true
+			},
+			createProjectsAndRulesGen,
```

Comment:
How to read a property-based test. Step 2: And how does it know how to generate something of the right shape? The second argument to the `func`, which is way down here. 😁